### ISSUE TITLE
Scene-graphs considered harmful

### ISSUE THREAD

**Issue by 3b** [bufferswap/ViralityEngine#93]
responding to some things in #90:
> One consequence of having this sort of model is that the scene tree is always a
single tree in the engine where each actor or component is tagged with one or
more domain tags that represent node membership in a conceptual domain.

and
> Hence, any API which allows querying for scene tree objects ...

I argue that "scene tree" should be replaced with "scene" when thinking about these things, and "the hierarchy of transformations" implied by "scene tree" should be considered as just another of the possible properties of the scene we might want to query.

If "querying ... tree objects" implies actually walking the tree, then thinking in terms of a "tree" abstraction is limiting our thinking and we might be missing optimizations we would have seen if we just had an arbitrary collection to query. If it doesn't imply walking the tree, then the tree wasn't the fundamental structure in the first place.

tl;dr notes:

* the "scene" is just an arbitrary collection
* objects are assumed static (and thus transforms can be baked, and they don't need to interact with the "transform hierarchy" used for updating things that move in local coordinate spaces) unless marked dynamic by app dev
* (`marked as dynamic` probably has a few forms, including `animated`, `physics system properties`, `attachment point`, or just a generic flag for things the dev wants to control from code)
* The `transform hierarchy` (aka `scene tree`) for things that do need it doesn't actually need to be an explicit tree data structure. Nodes might still store parent/child references, but we don't have to use those when doing the actual work that needs the hierarchy. It might be better to split it into separate structures for various purposes, for example updating skeletal animation nodes on GPU, but rigid body physics on CPU, each with their own optimized organization.

various notes from irc discussion:

* An `implicit scene graph` always exists: If there are multiple objects at distinct positions, and in particular if there are objects in local coordinate systems relative to other objects, we can always build a tree representing those relations, even if it ends up completely flat in some cases. 
* A potentially large subset of the `implicit scene graph` doesn't actually really need to be there: Children of the root node that never move would just be slowing down anything that does need to interact with the interesting parts of the scene hierarchy (and if not, they probably aren't using the scene graph data structure, and it wasn't really an important data structure to start with.)
    * `permanent static objects`: many games will have a lot of static geometry (the ground, buildings, furniture, etc), so having it all hanging off a root node of some data structure that thought it was a tree might be less efficient than just dumping it into a separate vector. 
    * `transient static objects`: sometimes we have an ability to place or remove objects, but they can't move once placed, and otherwise behave like permanent static objects.
    * alternately, distinguish by whether the object could be baked to an identity transform or not. A lot of static world geometry could do so if unique, but anything instanced or placed at runtime would need to retain a transform per instance.

* Interesting subtrees of the `implicit scene graph` fall into a few categories, and it might be worth storing them separately rather than in a single `explicit scene tree` data structure:
    * `physics objects`: another large group of objects like simple rigid-body physics objects will also frequently live directly under the root node, and separating those from the static objects again might be beneficial.
    * `simple animated/programmed objects`: (actually a subset of the `animated objects` below) Some objects like doors/buttons in a 3d game, or lots of things in a 2d game, might have simple enough behavior that handling them as "an object that moves relative to a node that exists only for that object to move around" might be  more than is needed.
    * things that actually have hierarchies: a few groups, again mostly living at the root, and then managing the children specially
        * `skeletons`: unusual in that they have multiple 'tree' nodes for a single (or small number) of objects
        * `composite physics objects`: cars with wheels, etc, where we have 1 main transform that is probably in world space, and some components that move relative to that due to physics (somewhat overlaps with `animated objects`, but likely to involve custom or at least specialized physics systems so listed separately)
        * `animated composite objects`: similarly, a ferris wheel probably will need to keep its hierarchy since components are animated in local spaces. (also vehicles might have animated or program-controlled nodes like propellers or doors)
        * `attachment points`: can be combined with any of the other cases (including static geometry). We frequently want to hold weapons, put skeletons on vehicles, etc, so will need to maintain (and update if applicable) the nodes for those attachment points.

* other views of the scene that are arguably as important as the "hierarchy of transforms" view:
    * `partition by materials`: (actually a tree of partitions, for example first by shader, then by textures, then by uniforms, or similar). Avoiding state changes helps rendering efficiency, so grouping things by state might be better.
    * `by distance from camera`: Similarly, some times we need to draw things in a particular order, so maybe a data structure optimized for that would be better. (Possibility of multiple cameras makes it more obvious that this one isn't *the* scene data structure, but doesn't make the other options better)
    * `prefab hierarchy`, `editing groups` etc: These frequently overlap with the `transform hierarchy`, but I'd argue they shouldn't be confused. For example the behavior you get when you think of "i dropped the orange on the plate on the table" might be different from what you get from "the orange is a child of the plate is a child of the table".  In the first case i might want an option to move all 3 as a group, but then if i turn the table sideways the plate and orange would still fall off once the move is done. If i turn the table in the second case, does the orange stay stuck to the plate and table? or does it roll off, but then continue to move with the table it is no longer touching? or was the child relationship an implementation detail, and i really wanted the first interpretation.        
Similarly "ownership of nested prefabs" probably tends to coincide with the transforms of the nested prefabs, but we should be able to bake in the transforms and keep the prefab relations for example.
    * alternate forms of the transform hierarchy: If we don't think of the `scene graph` as the main data structure, it is easier to split out the transforms from the objects, for example storing some or all of them in a big flat vector rather than with the object nodes might allow some optimizations for things that need to update a lot of them at once. For example putting things on the GPU, or improving cache locality on the CPU, or just being able to upload them all to GPU at once.

* benefits of change of perspective:
    * if we start from thinking "we have a tree of transformation relations containing the entire scene", then suddenly we have a bunch of redundant calculations when things didn't move, and we need to optimize those away, etc.
    * If we explicitly worry about those transformation relations as needed, a lot of those optimizations become less important, or go away completely. The nodes in the "currently playing animations" or "currently interacting physics objects" sets probably moved, and things in the "completely static objects" didn't. 
    * if we start with "the scene is a tree", we tend to think of operations in terms of "walking the tree"
    * if we start with "the scene is an arbitrary collection, + specialized index,view,etc", we would instead think "what is the best view for this operation", which probably isn't the transform hierarchy (possibly even in the case of updating the transforms).

**Comment by psilord** [comment]
This is excellent. My only comment is that an `attachment point` is often just a coordinate frame which may or may not be dynamic wrt its parent. Sometimes this `attachment point` might put the attaching object into multiple views. Such as the parent is both the coordinate parental frame AND ALSO the "owner" of the object, etc.  It is possible, too, that "attaching" an object to another might simply mean the attachee "owns" the attacher. Since we already use the term *attach* in V's code, we need to carefully build a vocabulary around this which probably involves the word *own*.

**Comment by 3b** [comment]
`attachment point` in that was intended specifically as "node in transform hierarchy onto which other subtrees of the transform hierarchy may be reparented at runtime (and which exist primarily for that purpose)", and unrelated to any other meaning those words might currently have in V, if that wasn't clear. That does sound like another of those cases where unrelated relationships could get confused with `transform hierarchy` and possibly shouldn't, though.

**Comment by psilord** [comment]
Also it is the case today in V that the scene objects are held in a bag as a set hash tables already with a pile of views into it for other things, so diminishing the importance of the scene tree in the code into a simple view of a particular relationship between actors should be not too painful. We do have to fix stuff like finding components and actors to not simply just walk the scene tree but walk the bags holding those things instead.

**Comment by psilord** [comment]
The more I think about 3b's viewpoint change in how scenes are dealt with, the more I like it.
